clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -119,33 +116,259 @@ void BoundsAnalysis::CollectBoundsVars(const Expr *E, DeclSetTy &BoundsVars) {
+  // updated them.
+  if (EB->Gen[SuccEB->Block].count(V))
+    return;
+
+  const llvm::APInt Zero(Ctx.getTypeSize(Ctx.IntTy), 0);
+
+  const llvm::APInt DerefOffsetVal = !DerefOffset ? Zero :
+    dyn_cast<IntegerLiteral>(DerefOffset)->getValue();
+
+  const llvm::APInt UpperOffsetVal = !UpperOffset ? Zero :
+    dyn_cast<IntegerLiteral>(UpperOffset)->getValue();
+
+  // We cannot widen the bounds if the offset in the deref expr is less than
+  // the offset in the upper bounds expr. For example:
+  // _Nt_array_ptr<T> p : bounds(p, p + i + 2); // Upper Bounds Offset (UBO) = 2.
+
+  // if (*(p + i)) // Offset = 0 < UBO ==> no widening
+  //   if (*(p + i + 1)) // Offset = 1 < UBO ==> no widening
+  //     if (*(p + i + 2)) // Offset = 2 == UBO ==> widen by 1
+  //       if (*(p + i + 3)) // Offset = 3 > UBO ==> widen by 2
+  if (Lex.CompareAPInt(DerefOffsetVal, UpperOffsetVal) ==
+      Lexicographic::Result::LessThan)
+    return;
+
+  EB->Gen[SuccEB->Block][V] =
+    (DerefOffsetVal - UpperOffsetVal).getLimitedValue();
+
+  // Any update to any variable involved in the bounds expr for the ntptr will
+  // kill the widenend bounds for the ntptr.
+  if (!SuccEB->BoundsVars.count(V)) {
+    DeclSetTy BoundsVars;
+    CollectBoundsVars(BE, BoundsVars);
+    SuccEB->BoundsVars[V] = BoundsVars;
+  }
+}
+
+ExprPairTy BoundsAnalysis::SplitIntoBaseOffset(const Expr *E) {
+  // In order to make an expression uniform, we want to keep all DeclRefExprs
+  // on the LHS and all IntegerLiterals on the RHS.
+
+  if (IsDeclOperand(E))
+    return std::make_pair(GetDeclOperand(E), nullptr);
+
+  if (!isa<BinaryOperator>(E))
+    return std::make_pair(nullptr, nullptr);
+
+  const BinaryOperator *BO = dyn_cast<BinaryOperator>(E);
+  Expr *LHS = BO->getLHS();
+  Expr *RHS = BO->getRHS();
+
+  if (isa<ParenExpr>(LHS))
+    LHS = IgnoreCasts(LHS);
+  if (isa<ParenExpr>(RHS))
+    RHS = IgnoreCasts(RHS);
+
+  // Note: Assume p, q, r are DeclRefExprs and i, j are IntegerLiterals.
+
+  // Case 1: LHS is DeclRefExpr and RHS is IntegerLiteral. This expr is already
+  // uniform.
+  // p + i ==> return (p, i)
+  if (IsDeclOperand(LHS) && isa<IntegerLiteral>(RHS))
+    return std::make_pair(GetDeclOperand(LHS), RHS);
+
+  // Case 2: LHS is IntegerLiteral and RHS is DeclRefExpr. We simply need to
+  // swap LHS and RHS to make expr uniform.
+  // i + p ==> return (p, i)
+  if (isa<IntegerLiteral>(LHS) && IsDeclOperand(RHS))
+    return std::make_pair(GetDeclOperand(RHS), LHS);
+
+  // Case 3: LHS and RHS are both DeclRefExpr's. This means there is no
+  // IntegerLiteral in the expr. In this case, we return the incoming
+  // BinaryOperator expr with a nullptr for the RHS.
+  // p + q ==> return (p + q, nullptr)
+  if (IsDeclOperand(LHS) && IsDeclOperand(RHS))
+    return std::make_pair(BO, nullptr);
+
+  if (!isa<BinaryOperator>(LHS) &&
+      !isa<BinaryOperator>(RHS))
+    return std::make_pair(nullptr, nullptr);
```

Comment:
Can you just return the binary operator with a nullptr for length?   This would let you get rid of the prior statement testing just for DeclOperands.

Reply:
Will try this tomorrow.